File: examples/basic/basic.c
/* Propensity file for basic URDME example.
 *
 * Reaction X+Y <--> Z in a 3D sphere.
 *
 * S. Engblom 2017-02-19 (Revision)
 * S. Engblom 2013-03-27 (Minor revision)
 * P. Bauer 2013-03-20
 */

#include "propensities.h"

/* number of reactions */
const int NR = 2;

/* forward declarations */
double rFun1(const int *x, double t, double vol,
	     const double *ldata, const double *gdata,int sd);
double rFun2(const int *x, double t, double vol,
	     const double *ldata, const double *gdata,int sd);

/* static propensity vector */
static PropensityFun ptr[] = {rFun1,rFun2};

/*----------------------------------------------------------------------*/
double rFun1(const int *x, double t, double vol,
	     const double *ldata, const double *gdata, int sd)
/* X+Y --> Z */
{
  return x[0]*x[1]/vol; 
}
/*----------------------------------------------------------------------*/
double rFun2(const int *x, double t, double vol,
	     const double *ldata, const double *gdata, int sd)
/* Z --> X+Y */
{
  return x[2];
}
/*----------------------------------------------------------------------*/
PropensityFun *ALLOC_propensities(size_t Mreactions)
{
  if (Mreactions > NR)
    mexErrMsgTxt("Wrong number of reactions.");
  return ptr;
}
/*----------------------------------------------------------------------*/
void FREE_propensities(PropensityFun* ptr)
{
  /* do nothing */
}
/*----------------------------------------------------------------------*/

File: examples/basic/basic.m
function umod = basic(umod)
%BASIC Model file for basic URDME example.

% S. Engblom 2013-03-27 (Minor revision)
% P. Bauer 2013-03-20

% sizes of problem
Mspecies = 3; % ordering is [X Y Z]
Ncells = numel(umod.vol); 

% stochiometric matrix N in sparse format
umod.N = sparse([-1  -1  1; ... % X+Y --> Z     
                  1   1 -1]');  % Z --> X+Y

% dependency graph G in sparse format
umod.G = sparse([1 1 0 1 1; ...
                 0 0 1 1 1]);
% 3 first columns correspond to diffusion events,
% 2 following columns to reactions...
% ...and the 2 rows are the two reactions

% initial number of species 
umod.u0 = zeros(Mspecies,Ncells); 

% distribute 50 X and 50 Y randomly
rng(123); % to get reproducible results
for i = 1:50
  cell = randi(Ncells,1,1);
  umod.u0(1,cell) = umod.u0(1,cell)+1;

  cell = randi(Ncells,1,1);
  umod.u0(2,cell) = umod.u0(2,cell)+1;
end

% output solution at times tspan
umod.tspan = 0:.1:100;
